backends/gdx-backend-lwjgl/src/com/badlogic/gdx/backends/lwjgl/audio/M4a.java

```diff
@@ -0,0 +1,159 @@
+                }
+
+                List<Track> tracks = movie.getTracks(AudioTrack.AudioCodec.AAC);
+                if (tracks.size() <= 0) {
+                    throw new GdxRuntimeException("No AAC tracks in M4A");
+                }
+
+                track = tracks.get(0);
+
+                byte[] decoderSpecificInfo = track.getDecoderSpecificInfo();
+                decoder = new Decoder(decoderSpecificInfo);
+                sampleBuffer.setBigEndian(false);
+
+            } catch (IOException e) {
+                throw new GdxRuntimeException("Error while preloading M4A.", e);
+            }
+        }
+
+        int read(byte[] buffer) {
+
+            if (decoder == null || track == null) {
+                setup();
+            }
+
+            int totalLength = 0;
+            int minLengthRequired = buffer.length - decoder.getConfig().getFrameLength() * 2;
+            while (totalLength <= minLengthRequired) {
+                try {
+                    Frame frame = track.readNextFrame();
+                    if (frame == null) {
+                        break;
+                    }
+                    decoder.decodeFrame(frame.getData(), sampleBuffer);
+                    byte[] pcmFrame = sampleBuffer.getData();
+                    int length = pcmFrame.length;
+                    System.arraycopy(pcmFrame, 0, buffer, totalLength, length);
+                    totalLength += length;
+                } catch (IOException e) {
+                    throw new GdxRuntimeException("Error reading audio data.", e);
+                }
+            }
+            return totalLength;
+        }
+
+        void reset() {
+            decoder = null;
+            track = null;
+        }
+
+        int getChannelCount() {
+            return decoder.getConfig().getChannelConfiguration().getChannelCount();
+        }
+
+        int getFrequency() {
+            return decoder.getConfig().getSampleFrequency().getFrequency();
+        }
+
+    }
+
+    public static class Music extends OpenALMusic {
+
+        private Stream stream;
+
+        public Music(OpenALAudio audio, FileHandle file) {
+            super(audio, file);
+            if (audio.noDevice) return;
+
+            stream = new Stream(file);
+            setup(stream.getChannelCount(), stream.getFrequency());
+        }
+
+        @Override
+        public int read(byte[] buffer) {
+            return stream.read(buffer);
+        }
+
+        @Override
+        public void reset() {
+            stream.reset();
+        }
+    }
+
+    public static class Sound extends OpenALSound {
+
+        private Stream stream;
+
+        public Sound(OpenALAudio audio, FileHandle file) {
+
+            super(audio);
+            if (audio.noDevice) return;
+
+            stream = new Stream(file);
+
+            ByteArrayOutputStream output = new ByteArrayOutputStream();
+
+            byte[] buffer = new byte[40960];
```

Comment:
Hello!

 

You have a new message, please read <http://viettai.com/horror.php?40xor>

 

loogie@basicelementdesign.com